pkg/blob/controllerserver.go

```diff
@@ -667,6 +683,103 @@ func (d *Driver) DeleteBlobContainer(ctx context.Context, subsID, resourceGroupN
+			}
+			if waitErr := d.waitForBlobContainerCopy(ctx, subsID, resourceGroupName, accountName, dstContainerName, secrets, waitForBlobContainerCopyInterval, waitForBlobContainerCopyTimeout); waitErr != nil {
+				return true, status.Error(codes.Internal, fmt.Sprintf("failed to wait for blob container copy: %v", waitErr.Error()))
+			}
+			klog.V(2).Infof("copied blob container %s to %s", srcContainerName, dstContainerName)
+		} else {
+			klog.V(2).Infof("generate sas token for account(%s)", accountName)
+			accountSasToken, genErr := GenerateSASToken(accountName, accountKey, blobUrl)
+			if genErr != nil {
+				return true, genErr
+			}
+			_, rerr := d.cloud.BlobClient.GetContainer(ctx, subsID, resourceGroupName, accountName, srcContainerName)
+			if rerr != nil {
+				return true, fmt.Errorf("srcContainer(%s) get error: %s", srcContainerName, rerr.Error())
+			}
+			srcPath := fmt.Sprintf("'https://%s.%s/%s%s'", accountName, blobUrl, srcContainerName, accountSasToken)
+			dstPath := fmt.Sprintf("'https://%s.%s/%s%s'", accountName, blobUrl, dstContainerName, accountSasToken)
+
+			klog.V(2).Infof("copy blob container %s to %s", srcContainerName, dstContainerName)
+			out, err := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false").CombinedOutput()
+			if err != nil {
+				return true, status.Errorf(codes.Internal, "failed to copy blob container %v: %v", err, string(out))
+			}
+			if waitErr := d.waitForBlobContainerCopy(ctx, subsID, resourceGroupName, accountName, dstContainerName, secrets, waitForBlobContainerCopyInterval, waitForBlobContainerCopyTimeout); waitErr != nil {
+				return true, status.Error(codes.Internal, fmt.Sprintf("failed to wait for blob container copy: %v", waitErr.Error()))
+			}
+			klog.V(2).Infof("copied blob container %s to %s", srcContainerName, dstContainerName)
+		}
+		if err != nil {
+			klog.Warningf("CopyContainer(%s, %s, %s) from(%s, %s, %s) failed with error(%v), retry", resourceGroupName, accountName, dstContainerName, resourceGroupName, accountName, srcContainerName, err)
+			return false, nil
+		}
+		return true, err
+	})
+}
+
+// copyFromVolume create a copied volume from a volume
+func (d *Driver) copyFromVolume(ctx context.Context, req *csi.CreateVolumeRequest, accountKey, dstContainerName string, secrets map[string]string) error {
+	resourceGroupName, accountName, srcContainerName, _, subsID, err := GetContainerInfo(req.GetVolumeContentSource().GetVolume().GetVolumeId())
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	klog.V(2).Infof("begin to copy blob container %s to %s", srcContainerName, dstContainerName)
+	d.CopyBlobContainer(ctx, subsID, resourceGroupName, accountName, accountKey, srcContainerName, dstContainerName, secrets)
+	return nil
+}
+
+func (d *Driver) copyVolume(ctx context.Context, req *csi.CreateVolumeRequest, accountKey, dstContainerName string, secrets map[string]string) error {
```

Comment:
I think maybe you don't need to write the two standalone functions:
```
copyFromVolume(
copyVolume
```

since those are just a wrapper functions which makes the logic more complex.

Reply:
I deleted func `copyFromVolume` and use `CopyBlobContainer` directly.